tests/topotests/bgp_listen_on_multiple_addresses/test_bgp_listen_on_multiple_addresses.py

```diff
@@ -0,0 +1,202 @@
+    tgen = get_topogen()
+    tgen.stop_topology()
+
+
+def test_startup():
+    """
+    Checks if bgpd starts properly.
+    """
+    tgen = get_topogen()
+    if tgen.routers_have_failure():
+        pytest.skip(tgen.errors)
+
+    # tgen.mininet_cli()
+
+    for router_name in tgen.routers().keys():
+        _wait_for_bgpd(router_name)
+
+
+def test_peering():
+    "Checks if the routers peer-up."
+    tgen = get_topogen()
+    if tgen.routers_have_failure():
+        pytest.skip(tgen.errors)
+
+    _wait_for_peer("r1", "10.0.0.2")
+    _wait_for_peer("r2", "10.0.0.1")
+    _wait_for_peer("r2", "10.0.1.2")
+    _wait_for_peer("r3", "10.0.1.1")
+    _wait_for_peer("r3", "10.0.2.2")
+    _wait_for_peer("r4", "10.0.2.1")
+
+
+def test_listening_address():
+    """
+    Checks if bgpd is only listening on the specified IP addresses.
+    """
+    tgen = get_topogen()
+    if tgen.routers_have_failure():
+        pytest.skip(tgen.errors)
+
+    for router in tgen.routers().values():
+        # bgpd must not be listening on the default address.
+        output = router.run("netstat -nlt4 | grep 0.0.0.0:179")
+        assert output == "", "{}: bpgd is listening on 0.0.0.0:179".format(router.name)
+
+        # bgpd must be listening on the specified addresses.
+        for address in LISTEN_ADDRESSES[router.name]:
+            output = router.run("netstat -nlt4 | grep {}:179".format(address))
+            assert output != "", "{}: bpgd not is listening on {}:179".format(
+                router.name, address
+            )
+
+
+def _wait_for_bgpd(router_name, timeout=60):
+    """
+    Waits for bgpd to start on a given router. If bgpd is not running on the
+    router within a given timeout (in seconds), then an exception is raised.
+    """
+    bgpd_ready = False
+    start_time = time.time()
+    tgen = get_topogen()
+    router = tgen.net[router_name]
+
+    while not bgpd_ready and time.time() - start_time < timeout:
+        bgpd_ready = "bgpd" in router.listDaemons()
+        if not bgpd_ready:
+            sleep(5, "{}: waiting for bgpd".format(router_name))
+
+    if not bgpd_ready:
+        raise Exception("{}: failed to start bgpd".format(router_name))
+
+
+def _wait_for_peer(router_name, peer, timeout=180):
+    """
+    Waits for the BGP connection between a given router and a given peer
+    (specified by its IP address) to be established. If the connection is
+    not established within a given timeout, then an exception is raised. 
+    """
+    peer_ready = False
+    start_time = time.time()
+    tgen = get_topogen()
+    router = tgen.routers()[router_name]
+
+    while not peer_ready:
+        try:
+            summary = router.vtysh_cmd("show ip bgp summary json", isjson=True)
+            assert summary["ipv4Unicast"]["peers"][peer]["state"] == "Established"
+        except:
+            if time.time() - start_time < timeout:
+                sleep(5, "{}: waiting for peer {}".format(router_name, peer))
```

Comment:
Could you change all _sleep_ stuff into something like?
```
    def _bgp_converge_initial():
        output = json.loads(
            tgen.gears[router].vtysh_cmd("show ip bgp summary json")
        )
        expected = {
            peer: {
                "bgpState": "Established"
            }
        }
        return topotest.json_cmp(output, expected)
```
Just an example (not tested with your changes).

Reply:
Revisiting the implementation was a good idea indeed. I ended up finding an even better way of implementing it: by using run_and_expect combined with router_json_cmp. I also figured that test_startup is not necessary, as routers_have_failures already verifies if bgpd is running.